# Kryptografia z kluczem tajnym (symetryczna): szyfry strumieniowe
### Literatura:
1. Nowoczesna kryptografia, Aumasson
2. Kryptografia dla praktyków, Schneier

### Historia szyfrów strumieniowych
1. Zapoznaj się z działaniem szyfrem Vernama

#### Funkcje pomocnicze
Zadaniem poniższego zestawu funkcji jest zamiana wiadomości tekstowej (poprzez kody ASCII) do (tekstowego) ciągu 0 i 1. Chcemy analizować funkcje tak jak są one opisane w książkach.


In [41]:
# Funkcje pomocnicze
from textwrap import wrap
import math
from codecs import encode

#tablica znaków w tablicę kodów int
def intoASCIIArray(message: str):
    int_array = []
    mesg_array = list(message)
    for i in mesg_array:
        int_array.append(ord(i))
    return int_array

#tablica kodów int w tablice znaków
def intoCharArray(message: []):
    mesg_char = []
    for i in message:
        mesg_char.append(chr(i))
    return mesg_char

# jak wyświetlić dane w postaci binarnej na n bitach
get_bin = lambda x, n: format(x, 'b').zfill(n)
get_dec = lambda x: int(x,2)
#tekst ascii w postaci tablicy 8-bitowych porcji
def ASCIIToBinChunks(message_list):
    binary = []
    for x in message_list:
        binary.append(get_bin(x, 8))
    return binary

#tekst ascii w formie strumienia bitów
def ASCIIToBinStream(binary: []):
    binary_str = ""
    for x in binary:
        binary_str+=x
    return binary_str


# Przykłady

# Liczba 200 w zapisie binarnym
l = 200
bl = get_bin(l,8)
print("Liczba 200 w zapisie binarnym na 8 bitach:", bl)
bl = get_bin(l,16)
print("Liczba 200 w zapisie binarnym na 16 bitach:", bl)


messageTxt = 'The quick brown fox jumps over the lazy dog'
messageASCII = intoASCIIArray(messageTxt)
print("Tekst w formie kodów ASCII: ",messageASCII)
messageASCIIBinChunks = ASCIIToBinChunks(messageASCII)
print("Tekst w formie porcji bitów:", messageASCIIBinChunks)
messageASCIIBinStream = ASCIIToBinStream(messageASCIIBinChunks)
print("Tekst w postaci ciągu 0 i 1:", messageASCIIBinStream)
print()


#Operatory logiczne działające na napisach bitów
def XOR(bits1,bits2):
    """perform a XOR operation and return the output"""
    xor_result = ""
    for index in range(len(bits1)):
        if bits1[index] == bits2[index]:
            xor_result += '0'
        else:
            xor_result += '1'
    return xor_result

def AND(bits1,bits2):
    """perform a AND operation and return the output"""
    and_result = ""
    for index in range(len(bits1)):
        if (bits1[index] == '1') and  (bits2[index] == '1'):
            and_result += '1'
        else:
            and_result += '0'
    return and_result

def OR(bits1,bits2):
    """perform a OR operation and return the output"""
    or_result = ""
    for index in range(len(bits1)):
        if (bits1[index] == '0') and  (bits2[index] == '0'):
            or_result += '0'
        else:
            or_result += '1'
    return or_result

def NEG(bits):
    """perform a NEG operation and return the output"""
    neg_result = ""
    for index in range(len(bits)):
        if (bits[index] == '0'):
            neg_result += '1'
        else:
            neg_result += '0'
    return neg_result

print("AND", AND('0101010001101', '0101010001100'))
print("OR", OR('0101010001101', '0101010001100'))
print("XOR", XOR('0101010001101', '0101010001100'))
print("NEG", NEG('0101010001101'))

Liczba 200 w zapisie binarnym na 8 bitach: 11001000
Liczba 200 w zapisie binarnym na 16 bitach: 0000000011001000
Tekst w formie kodów ASCII:  [84, 104, 101, 32, 113, 117, 105, 99, 107, 32, 98, 114, 111, 119, 110, 32, 102, 111, 120, 32, 106, 117, 109, 112, 115, 32, 111, 118, 101, 114, 32, 116, 104, 101, 32, 108, 97, 122, 121, 32, 100, 111, 103]
Tekst w formie porcji bitów: ['01010100', '01101000', '01100101', '00100000', '01110001', '01110101', '01101001', '01100011', '01101011', '00100000', '01100010', '01110010', '01101111', '01110111', '01101110', '00100000', '01100110', '01101111', '01111000', '00100000', '01101010', '01110101', '01101101', '01110000', '01110011', '00100000', '01101111', '01110110', '01100101', '01110010', '00100000', '01110100', '01101000', '01100101', '00100000', '01101100', '01100001', '01111010', '01111001', '00100000', '01100100', '01101111', '01100111']
Tekst w postaci ciągu 0 i 1: 0101010001101000011001010010000001110001011101010110100101100011011010110010000

In [40]:
# Generowanie losowych danych: funkcje i algorytmy pomocnicze
import random
random.seed("key")  #generator PRNG w python można inicjalizować tekstem

# jak wygenerować 8 losowych bitów (razy n)
def randomBytes(n):
    return bytes(random.getrandbits(8) for i in range(n))

def randomByte():
    return random.randint(0,255)

def randomByteBin():
    return get_bin(random.randint(0,255),8)

#jak wygenerować ciąg losowych bajtów
random_stream_chunks = []
for i in range(8):
    random_stream_chunks.append(randomByteBin())

print("Losowe bajty:", random_stream_chunks)

randomBinaryString = ""
for i in random_stream_chunks:
    randomBinaryString+=i

print("Strumień bitów:", randomBinaryString)

print("Pojedyncze losowe bajty:", end="")
bits1 = randomByteBin()
bits2 = get_bin(randomByte(),8)
print(bits1, bits2)

Losowe bajty: ['10001110', '11111000', '11000111', '11110101', '11101100', '01010100', '00111110', '10011110']
Strumień bitów: 1000111011111000110001111111010111101100010101000011111010011110
Pojedyncze losowe bajty:10000110 00000010


### Zadanie samodzielne dla studentów
Zaimplementuj szyfr, który utajnia strumień wiadomości jawnej przez łączenie go z pseudolosowym strumieniem


In [47]:
from math import pow
s='''Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc eget augue eget sem sodales ultrices. Quisque dapibus, urna sit amet.'''
# %store s >plaintext.txt
# h = open("plaintext.txt", "r")

random.seed("key") #przykład



# tu wpisz swoj kod
message_lenght = len(s)
same_key = [randomByteBin() for _ in range(message_lenght)]
random_BinStream= ASCIIToBinStream(same_key)



def messageToBinStream(text:str) -> str:
    messageASCII = intoASCIIArray(text)
    messageBinChunks = ASCIIToBinChunks(messageASCII)
    messageBinStream = ASCIIToBinStream(messageBinChunks)
    return messageBinStream

def vernam_ciphertext(plaintext:str, random_BinStream:list = None) -> str:
    plaintext = plaintext.lower()
    if random_BinStream == None:
        random_BinStream = [randomByteBin() for _ in range(len(plaintext))]
        random_BinStream= ASCIIToBinStream(random_BinStream)
    
    plaintext_BinStream = messageToBinStream(plaintext)


    ciphertext = XOR(plaintext_BinStream,random_BinStream)
    return ciphertext

def binInToDecimal(binary:str) -> int:
    power = 0
    decimal = 0
    while binary != "":
        binary





def vernam_decode(cipthertext_BinStream:str,random_BinSream:str,bytes_length:str=8) -> str:
    
    decoded_BinStream = XOR(cipthertext_BinStream,random_BinStream)
    decoded_Bytes = list(map(lambda x: decoded_BinStream[x:x+bytes_length], range(0, len(decoded_BinStream), bytes_length)))
    decoded_ASCII = list(map(lambda x: get_dec(x),decoded_Bytes))
    decoded_message = intoCharArray(decoded_ASCII)
    decodedtext = "".join(decoded_message)
    return decodedtext




print("Zaszyfrowany BitStream: ", vernam_ciphertext(s,random_BinStream))
print("Odszyfrowana wiadomość: ",  vernam_decode(vernam_ciphertext(s,random_BinStream),random_BinStream))

Zaszyfrowany BitStream:  111000101001011110110101100100001000000101110100010101111110111011110101011101111011111111111101001101000010010010001111011001101110101110001100110100101110010111010011110001110000011100100100111111000000011001101001010011010001011111110011000100100100100110111011000010000010100001101100101011110001111110101110010010101111100011010011011010101111110001001101110000111010111110000111111000110001110011011011010011001111111100110001101110110101001010100011010101010101000111111001010001011011001110101101101010101001010101110000000000010001101000000110000001100000110010010100001111010011001010101100010100100011001000101010100101010110110101100100111011011000010100110010000001100110101100011011010101001010110000100000000101010110011101110111101100110100010111011010000100111011101101110101000001010001011100101101001111110011001100101100101100001100110100010010001100100011110011010101100101000110001101010010001101101010100110101000001110001000111101001010001010110101010

### Zastanów się
1. W jaki sposób łączymy strumień tekstu jawnego i strumień klucza? Wady i zalety różnych metod.

### Linear congruential generators (LCG)
Generatory liniowe kongruentne sa najprostszymi generatorami dającymi ciągi liczb o dobrych własnościach statystycznych. Kolejne wyrazy ciągu generowane sa przy pomocy formuły:

$z_{i+1}=(a\cdot z_i+c)\bmod m$

Wyraz $z_0$ nazywany jest ziarnem (_seed_). Użycie tego samego ziarna gwarantuje nam wyprodukowanie tej samej sekwencji liczb.

Charakterystyczną cechą GLK jest ich okresowość. Oczekujemy możliwie najdłuższego okresu (maksymalny to $m-1$). Najdłuższy okres otrzymamy gdy spełnione są pewne warunki (twierdzenie Hull'a-Dobell'a):

- $c$ i $m$ są względnie pierwsze,
- $a-1$ jest podzielne przez wszystkiem pierwsze czynniki $m$,
- $a-1$ jest wielokrotnoścą 4 jeśli $m$ jest wielokrotnością 4.

Przykładowe dobre wartości to $a=1103515245$, $c=12345$ dla $m=2^{31}$

Zazwyczaj generator zwraca wartość $\frac{z_i}{m}$, ale wyjście można przeskalować do dowolnej innej wartości.

Obecnie większość PRNG to tzw. _Mersenne twister_, ale ogólna idea ich użytkowania i własności jest taka sama jak w przypadku generatorów kongruentnych.

In [48]:
def lcg(x, a, c, m):
    while True:
        x = (a * x + c) % m
        yield x

def random_uniform_sample(n, interval, seed=0):
    a, c, m = 1103515245, 12345, 2 ** 31
    bsdrand = lcg(seed, a, c, m)

    lower, upper = interval[0], interval[1]
    sample = []

    for i in range(n):
        observation = (upper - lower) * (next(bsdrand) / (2 ** 31 - 1)) + lower
        sample.append(round(observation))

    return sample

print(random_uniform_sample(100, [0,255],11))

[166, 97, 88, 33, 91, 184, 211, 119, 189, 205, 46, 8, 82, 18, 98, 253, 139, 27, 203, 49, 203, 44, 215, 76, 5, 204, 67, 84, 70, 232, 23, 190, 31, 114, 226, 213, 33, 239, 96, 162, 104, 13, 14, 12, 112, 126, 219, 108, 23, 86, 124, 87, 47, 36, 130, 187, 97, 233, 206, 39, 210, 88, 112, 86, 72, 72, 144, 215, 63, 238, 132, 236, 182, 191, 142, 108, 23, 115, 119, 207, 4, 53, 145, 53, 114, 93, 60, 125, 219, 199, 126, 74, 236, 39, 212, 15, 121, 91, 202, 235]


### Zastanów sie
1. Dlaczego stosowanie generatorów PRNG jako źródłą strumienia klucza to zły pomysł?

### Kryptograficzne generatory PRNG
Urządzenie /dev/urandom stanowi podstawę dobrego generatora CPRNG

In [ ]:
import os
import struct

# random integer using os.urandom()
print(struct.unpack('i', os.urandom(4)))
# Output (258871565,)

# unsigned random integer using os.urandom()
print(struct.unpack('I', os.urandom(4)))
print(struct.unpack('I', os.urandom(4))[0] % 100)
# Output (1015967885,)

# random short number using os.urandom()
print(struct.unpack('h', os.urandom(2)))
# Output (-28882,)

# unsigned random short using os.urandom()
print(struct.unpack('H', os.urandom(2)))
# Output (29492,)

# Print random float using os.urandom()
print(struct.unpack('f', os.urandom(4)))
# Output (-4.651611836498911e+20,)

# un-singed random decimal using os.urandom()
print(struct.unpack('d', os.urandom(8)))
# Output (-1.7024488468332834e-120,)

# random char using os.urandom()
print(struct.unpack('c', os.urandom(1)))
# Output (b'\xce',)

### Zastanów się:
1. Poszukaj informacji o szyfrach binarnie addytywnych
2. Poszukaj informacji o szyfrach strumieniowych używanych w praktyce. Gdzie takie szyfry mogą być obecnie stosowane?

### Problem
Utwórz dwie różne wiadomości równej długości. Zaszyfruj je szyfrem XOR z użyciem tego samego klucza. Wyznacz alternatywę rozłączną szyfrogramów (XOR) i porównaj ją z tą samą operacją wykonaną dla tekstów jawnych.  

In [65]:
message1 = 'secret'
message2 = 'poufny'
key = "coffee"
samekey_BinStream= messageToBinStream(key)

message1_BinStream = messageToBinStream(message1)
message2_BinStream = messageToBinStream(message2)
print("Message 1 BinStream",message1_BinStream)
print("Message 2 BinStream",message2_BinStream)


message1_coded = vernam_ciphertext(message1,samekey_BinStream)
message2_coded = vernam_ciphertext(message2,samekey_BinStream)
print("Message 1 coded",message1_coded)
print("Message 2 coded",message2_coded)



print("XOR messeges BinStream",XOR(message1_BinStream,message2_BinStream))
print("XOR messeges coded",XOR(message1_coded,message2_coded))
# same result
print(XOR(message1_coded,message2_coded)==XOR(message1_BinStream,message2_BinStream))







Message 1 BinStream 011100110110010101100011011100100110010101110100
Message 2 BinStream 011100000110111101110101011001100110111001111001
Message 1 coded 000100000000101000000101000101000000000000010001
Message 2 coded 000100110000000000010011000000000000101100011100
XOR messeges BinStream 000000110000101000010110000101000000101100001101
XOR messeges coded 000000110000101000010110000101000000101100001101
True


### Bezpieczeństwo szyfru XOR
1. Jakie znaczenie ma powyższy wynik z punktu widzenia kryptoanalizy?
2. Jeśli OTP to OK.
3. Na czym polega atak ze znanym tekstem jawnym?

### Problem

1. Utwórz dowolną wiadomość $M_1$.
2. Zaszyfruj ją swoim szyfrem XOR z kluczem $K$.
3. Wykonaj na szyfrogramie $C_1$ operację $C_2 = C_1 \oplus (111\ldots1)$.
4. Odszyfruj wiadomość $C_2$ stosując ten sam klucz $K$.
5. Porównaj wiadomości: $M_1$ i odszyfrowaną w poprzednim kroku $M_2$ (najlepiej binarnie).  